packages/core/src/routes/session/utils.ts

```diff
@@ -32,27 +41,34 @@ export const getPasswordlessRelatedLogType = (
   return result.data;
 };
 
-export const parseVerificationStorage = (data: unknown): VerificationStorage => {
-  const verificationResult = z
-    .object({
-      verification: verificationStorageGuard,
-    })
-    .safeParse(data);
+const parseVerificationStorage = (data: unknown): VerificationStorage => {
+  const verificationResult = verificationResultGuard.safeParse(data);
 
-  assertThat(
-    verificationResult.success,
-    new RequestError({
-      code: 'session.verification_session_not_found',
-      status: 404,
-    })
-  );
+  if (!verificationResult.success) {
+    throw new RequestError(
+      {
+        code: 'session.verification_session_not_found',
+        status: 404,
+      },
+      verificationResult.error
+    );
+  }
 
   return verificationResult.data.verification;
 };
 
-export const verificationSessionCheckByFlow = (
+export const getVerificationStorageFromInteraction = async (
+  ctx: Context,
+  provider: Provider
+): Promise<VerificationStorage> => {
+  const { result } = await provider.interactionDetails(ctx.req, ctx.res);
+
+  return parseVerificationStorage(result);
+};
+
+export const checkVerificationSessionByFlow = (
   currentFlow: PasscodeType,
-  payload: Pick<VerificationStorage, 'flow' | 'expiresAt'>
+  payload: VerificationStorage
```

Comment:
Actually, I mean merge all these into four different zod guards.  So you don't need all the separate checks.

e.g.
```js
const smsSignInSessionGuard = z.object({
 flow: z.literal(PasscodeType.SignIn),
 phone: z.string(),
 expiresAt: z.string(),
});

type SmsSignInSessionStorage = zod.infer(smsSignInSessionGuard):

const getVerificationStorageFromInteraction = (ctx, provider, sessionGuard) => async {
  const { result } = await provider.interactionDetails(ctx.req, ctx.res);
  const storage = sessionGuard.safeParse(result);

  if (!storage.success) {
    throw new RequestError(
      {
        code: 'session.verification_session_not_found',
        status: 404,
      },
      verificationResult.error
    );
  }
  
  return storage.data.
}

```

So you will get a well typed and guarded result.

Reply:
It's up to you. Either one works for me.